### PR TITLE
App registration - configuring apps

### PR DESCRIPTION
**Pull request by monikasulik**
2nd and 3rd subtask of app registration combined. For internal review.

### CODE REVIEW

**Comment by Aiky30** [comment]
Please confirm that this is the correct request for me to look at as I know you had one previously that you told me to ignore. Ta.

**Comment by Aiky30** [comment]
One thing that I think as a whole team we should be considering is roughly documenting any new features that we're creating, when it comes to working with the person creating the docs it will make it much easier.

File: cms/tests/test_app_registration.py [hunk 1]
```diff
@@ -53,4 +53,186 @@ def test_imports_cms_apps_files(self):
     ])
     def test_raises_exception_raised_in_cms_file(self):
```
Comment (Aiky30) [inline cms/tests/test_app_registration.py:54]: Should be a new line here. Picky I know. :-)

Comment (Aiky30) [inline cms/tests/test_app_registration.py:54]: Should be a new empty line here. Picky I know. :-)

File: cms/tests/test_app_registration.py [hunk 2]
```diff
@@ -53,4 +53,186 @@ def test_imports_cms_apps_files(self):
...
         with self.assertRaises(KeyError):
-            autodiscover_cms_files()
+            app_registration.autodiscover_cms_configs()
+
+    @override_settings(INSTALLED_APPS=[
+        'cms.tests.test_app_registry.app_with_cms_extension',
+        'cms.tests.test_app_registry.app_without_cms_file'
+    ])
+    def test_adds_cms_app_attribute_to_django_app_config(self):
```
Comment (Aiky30) [inline cms/tests/test_app_registration.py:62]: I would personally add a small comment on what each test is looking for no matter how obvious it seems to you. I had to look at it a few times to follow it.

File: cms/tests/test_app_registration.py [hunk 3]
```diff
@@ -53,4 +53,186 @@ def test_imports_cms_apps_files(self):
...
+        config_app_enabled.cms_app = Mock(
+            spec=['djangocms_feature_x_enabled'])
+        config_app_enabled.cms_app.djangocms_feature_x_enabled = True
+        # Pretend these mocked apps are in INSTALLED_APPS
+        mocked_apps.return_value = [
+            feature_app, config_app_enabled]
+
+        app_registration.configure_cms_apps([feature_app])
+
+        feature_app.cms_app.configure_app.assert_called_once_with(
```
Comment (Aiky30) [inline cms/tests/test_app_registration.py:135]: What is this assertion doing? A small comment would make it easy for me to understand. Again personal opinion.

File: cms/app_registration.py [hunk 4]
```diff
@@ -1,27 +1,96 @@
...
         try:
-            import_module('%s.%s' % (app_config.name, CMS_CONFIG_NAME))
-        except Exception:
+            cms_module = import_module(
+                '%s.%s' % (app_config.name, CMS_CONFIG_NAME))
+        except:
+            # If something in cms_apps.py raises an exception let that
+            # exception bubble up. Only catch the exception if
+            # cms_apps.py doesn't exist
             if module_has_submodule(app_config.module, CMS_CONFIG_NAME):
```
Comment (Aiky30) [inline cms/app_registration.py:25]: Is this bubble up exception tested? I couldn't see a test that mentioned this.

File: cms/tests/test_app_registration.py [hunk 5]
```diff
@@ -53,4 +53,186 @@ def test_imports_cms_apps_files(self):
...
+        non_cms_app = Mock(spec=AppConfig)
+        # Pretend these mocked apps are in INSTALLED_APPS
+        mocked_apps.return_value = [
+            feature_app_x, non_cms_app, config_app_xy, config_app_y,
+            config_app_x, feature_app_y]
+
+        app_registration.configure_cms_apps(
+            [feature_app_x, feature_app_y])
+
+        # Assert we configured the 2 apps we expected with feature x
```
Comment (Aiky30) [inline cms/tests/test_app_registration.py:221]: This comment saved me having to process what was going on. Me like.:-)

File: cms/tests/test_app_registration.py [hunk 6]
```diff
@@ -53,4 +53,186 @@ def test_imports_cms_apps_files(self):
...
+        self.assertEqual(
+            feature_app_x.cms_app.configure_app.call_count, 2)
+        self.assertEqual(
+            feature_app_x.cms_app.configure_app.call_args_list[0][0][0],
+            config_app_xy)
+        self.assertEqual(
+            feature_app_x.cms_app.configure_app.call_args_list[1][0][0],
+            config_app_x)
+        # Assert we configured the 2 apps we expected with feature y
```
Comment (Aiky30) [inline cms/tests/test_app_registration.py:230]: This comment saved me having to process what was going on. Me like.:-)

File: cms/tests/test_app_registration.py [hunk 7]
```diff
@@ -53,4 +53,186 @@ def test_imports_cms_apps_files(self):
...
+        mocked_apps.return_value = [
+            app_with_cms_ext1,
+            app_with_cms_config,
+            app_with_cms_ext2,
+            non_cms_app,
+        ]
+
+        apps_with_features = app_registration.get_cms_apps_with_features()
+
+        self.assertListEqual(
```
Comment (Aiky30) [inline cms/tests/test_app_registration.py:109]: What is this assertion doing i.e. what is it testing for? A small comment would make it easy for me to understand. Again personal opinion.

**Review by Aiky30 — COMMENTED** [review]
Sorry that I cannot be more helpful with the design etc. I get what the tasks are for but I'm not involved enough to make any design comments so this review will just be me being picky. Paulo will be the man to make those comments I guess.

I would argue that comments on assertions are good so  Paulo can see what you were trying to test for (even if really obvious) because he might have other ideas of how to test for what the comment describes is supposed to be happening. It would also save me alot of time reviewing as i can follow what you're trying to test or rather than what you're actually testing for (well my interpretation of the test).

May I be excused to do some work now boss? :-P

**Comment by monikasulik** [comment]
@Aiky30 I will do what I think Paulo wanted which is document in detail in the PR description once I submit the whole thing :) I think the idea was that Daniele would later write docs based on the PR description.

File: cms/app_registration.py (hunk 4 above)
Comment (monikasulik) [reply cms/app_registration.py:25]: https://github.com/monikasulik/django-cms/blob/b5f1825e1225e00de25f6fe4114a512c04c18708/cms/tests/test_app_registration.py#L54
:smile:

File: cms/tests/test_app_registration.py (hunk 1 above)
Comment (monikasulik) [reply cms/tests/test_app_registration.py:54]: Being picky back :wink: AFAIK pep8 doesn't require it and I actually prefer it visually without. However, if there is a general convention in django-cms to add an empty line (I haven't noticed this, but feel free to direct me to examples) then I can keep to it for consistency.

**Comment by monikasulik** [comment]
Am merging this and will add the assert comments in the next PR :)

**Pull request merged**